File: mycelix-workspace/sdk-ts/src/utils/pagination.ts
// Copyright (C) 2024-2026 Tristan Stoltz / Luminous Dynamics
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial licensing: see COMMERCIAL_LICENSE.md at repository root
/**
 * @mycelix/sdk Pagination Utilities
 *
 * Standardized pagination for all Mycelix APIs.
 * Supports cursor-based and offset-based pagination.
 *
 * @packageDocumentation
 * @module utils/pagination
 */

/** Default page size when not specified */
export const DEFAULT_PAGE_SIZE = 50;

/** Maximum allowed page size to prevent resource exhaustion */
export const MAX_PAGE_SIZE = 500;

/**
 * Pagination request parameters
 */
export interface PaginationRequest {
  /** Number of items per page (capped at MAX_PAGE_SIZE) */
  pageSize?: number;
  /** Cursor for cursor-based pagination (opaque string) */
  cursor?: string;
  /** Offset for offset-based pagination */
  offset?: number;
  /** Sort direction (true = descending/newest first) */
  sortDesc?: boolean;
}

/**
 * Paginated response wrapper
 */
export interface PaginatedResponse<T> {
  /** Items in this page */
  items: T[];
  /** Total count of items (if available) */
  totalCount?: number;
  /** Cursor for next page (if cursor-based) */
  nextCursor?: string;
  /** Whether there are more items */
  hasMore: boolean;
  /** Current page number (if offset-based) */
  page?: number;
  /** Total pages (if offset-based and total known) */
  totalPages?: number;
}

/**
 * Create a pagination request with defaults
 */
export function createPaginationRequest(
  pageSize: number = DEFAULT_PAGE_SIZE,
  cursor?: string
): PaginationRequest {
  return {
    pageSize: Math.min(pageSize, MAX_PAGE_SIZE),
    cursor,
    sortDesc: true,
  };
}

/**
 * Create a pagination request with offset
 */
export function createOffsetPaginationRequest(
  pageSize: number = DEFAULT_PAGE_SIZE,
  offset: number = 0
): PaginationRequest {
  return {
    pageSize: Math.min(pageSize, MAX_PAGE_SIZE),
    offset,
    sortDesc: true,
  };
}

/**
 * Get effective page size (capped at MAX_PAGE_SIZE)
 */
export function getEffectivePageSize(request: PaginationRequest): number {
  return Math.min(request.pageSize ?? DEFAULT_PAGE_SIZE, MAX_PAGE_SIZE);
}

/**
 * Get effective offset
 */
export function getEffectiveOffset(request: PaginationRequest): number {
  return request.offset ?? 0;
}

/**
 * Paginate an array of items
 */
export function paginateArray<T>(
  items: T[],
  request: PaginationRequest
): PaginatedResponse<T> {
  const total = items.length;
  const offset = getEffectiveOffset(request);
  const pageSize = getEffectivePageSize(request);

  if (offset >= total) {
    return {
      items: [],
      totalCount: total,
      hasMore: false,
      page: 0,
      totalPages: Math.ceil(total / pageSize),
    };
  }

  const end = Math.min(offset + pageSize, total);
  const pageItems = items.slice(offset, end);
  const hasMore = end < total;

  return {
    items: pageItems,
    totalCount: total,
    nextCursor: hasMore ? end.toString() : undefined,
    hasMore,
    page: Math.floor(offset / pageSize),
    totalPages: Math.ceil(total / pageSize),
  };
}

/**
 * Create an empty paginated response
 */
export function emptyPaginatedResponse<T>(): PaginatedResponse<T> {
  return {
    items: [],
    totalCount: 0,
    hasMore: false,
    page: 0,
    totalPages: 0,
  };
}

/**
 * Cursor encoding/decoding utilities
 */
export const cursor = {
  /**
   * Encode a cursor from timestamp and optional ID
   */
  encode(timestampMicros: number, id?: string): string {
    return id ? `${timestampMicros}:${id}` : timestampMicros.toString();
  },

  /**
   * Decode a cursor to timestamp and optional ID
   */
  decode(cursor: string): { timestamp: number; id?: string } | null {
    if (cursor.includes(':')) {
      const [tsStr, ...idParts] = cursor.split(':');
      const timestamp = parseInt(tsStr, 10);
      if (isNaN(timestamp)) return null;
      return { timestamp, id: idParts.join(':') };
    }
    const timestamp = parseInt(cursor, 10);
    if (isNaN(timestamp)) return null;
    return { timestamp };
  },
};

/**
 * Pagination helper class for building paginated queries
 */
export class Paginator<T> {
  private items: T[] = [];
  private pageSize: number;
  private currentOffset: number = 0;

  constructor(pageSize: number = DEFAULT_PAGE_SIZE) {
    this.pageSize = Math.min(pageSize, MAX_PAGE_SIZE);
  }

  /**
   * Set items to paginate
   */
  setItems(items: T[]): this {
    this.items = items;
    return this;
  }

  /**
   * Get current page
   */
  getCurrentPage(): PaginatedResponse<T> {
    return paginateArray(this.items, {
      pageSize: this.pageSize,
      offset: this.currentOffset,
    });
  }

  /**
   * Go to next page
   */
  nextPage(): PaginatedResponse<T> | null {
    const current = this.getCurrentPage();
    if (!current.hasMore) return null;
    this.currentOffset += this.pageSize;
    return this.getCurrentPage();
  }

  /**
   * Go to previous page
   */
  prevPage(): PaginatedResponse<T> | null {
    if (this.currentOffset === 0) return null;
    this.currentOffset = Math.max(0, this.currentOffset - this.pageSize);
    return this.getCurrentPage();
  }

  /**
   * Go to specific page
   */
  goToPage(page: number): PaginatedResponse<T> {
    this.currentOffset = page * this.pageSize;
    return this.getCurrentPage();
  }

  /**
   * Reset to first page
   */
  reset(): this {
    this.currentOffset = 0;
    return this;
  }

  /**
   * Get all pages as an async iterator
   */
  async *pages(): AsyncGenerator<PaginatedResponse<T>> {
    this.reset();
    let page = this.getCurrentPage();
    yield page;

    while (page.hasMore) {
      const next = this.nextPage();
      if (!next) break;
      page = next;
      yield page;
    }
  }
}

/**
 * Create a paginator instance
 */
export function createPaginator<T>(pageSize?: number): Paginator<T> {
  return new Paginator<T>(pageSize);
}

/**
 * Pagination utilities for async data fetching
 */
export const asyncPagination = {
  /**
   * Fetch all pages and combine results
   */
  async fetchAll<T>(
    fetchFn: (request: PaginationRequest) => Promise<PaginatedResponse<T>>,
    initialRequest: PaginationRequest = { pageSize: DEFAULT_PAGE_SIZE }
  ): Promise<T[]> {
    const allItems: T[] = [];
    let request = { ...initialRequest };
    let hasMore = true;

    while (hasMore) {
      const response = await fetchFn(request);
      allItems.push(...response.items);
      hasMore = response.hasMore;

      if (hasMore && response.nextCursor) {
        request = { ...request, cursor: response.nextCursor };
      } else if (hasMore) {
        request = {
          ...request,
          offset: (request.offset ?? 0) + getEffectivePageSize(request),
        };
      }

      // Safety limit
      if (allItems.length > 10000) {
        console.warn('Pagination safety limit reached (10000 items)');
        break;
      }
    }

    return allItems;
  },

  /**
   * Create an async iterator for paginated results
   */
  async *iterate<T>(
    fetchFn: (request: PaginationRequest) => Promise<PaginatedResponse<T>>,
    initialRequest: PaginationRequest = { pageSize: DEFAULT_PAGE_SIZE }
  ): AsyncGenerator<T> {
    let request = { ...initialRequest };
    let hasMore = true;

    while (hasMore) {
      const response = await fetchFn(request);

      for (const item of response.items) {
        yield item;
      }

      hasMore = response.hasMore;

      if (hasMore && response.nextCursor) {
        request = { ...request, cursor: response.nextCursor };
      } else if (hasMore) {
        request = {
          ...request,
          offset: (request.offset ?? 0) + getEffectivePageSize(request),
        };
      }
    }
  },
};

File: mycelix-workspace/sdk/src/pagination.rs
// Copyright (C) 2024-2026 Tristan Stoltz / Luminous Dynamics
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial licensing: see COMMERCIAL_LICENSE.md at repository root
//! # Pagination Module
//!
//! Provides standardized pagination utilities for Mycelix zomes.
//! Supports cursor-based and offset-based pagination patterns.
//!
//! ## Example
//!
//! ```rust
//! use mycelix_sdk::pagination::{PaginationRequest, PaginatedResponse, paginate_vec};
//!
//! let items = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
//! let request = PaginationRequest::new(3, None); // Page size 3
//!
//! let response = paginate_vec(items, &request);
//! assert_eq!(response.items.len(), 3);
//! assert!(response.has_more);
//! ```

use serde::{Deserialize, Serialize};

/// Default page size when not specified
pub const DEFAULT_PAGE_SIZE: u32 = 50;

/// Maximum allowed page size to prevent resource exhaustion
pub const MAX_PAGE_SIZE: u32 = 500;

/// Pagination request parameters
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PaginationRequest {
    /// Number of items per page (capped at MAX_PAGE_SIZE)
    pub page_size: u32,
    /// Cursor for cursor-based pagination (opaque string)
    pub cursor: Option<String>,
    /// Offset for offset-based pagination
    pub offset: Option<u32>,
    /// Sort direction
    pub sort_desc: bool,
}

impl Default for PaginationRequest {
    fn default() -> Self {
        Self {
            page_size: DEFAULT_PAGE_SIZE,
            cursor: None,
            offset: None,
            sort_desc: true, // Most recent first by default
        }
    }
}

impl PaginationRequest {
    /// Create a new pagination request with page size
    pub fn new(page_size: u32, cursor: Option<String>) -> Self {
        Self {
            page_size: page_size.min(MAX_PAGE_SIZE),
            cursor,
            offset: None,
            sort_desc: true,
        }
    }

    /// Create with offset-based pagination
    pub fn with_offset(page_size: u32, offset: u32) -> Self {
        Self {
            page_size: page_size.min(MAX_PAGE_SIZE),
            cursor: None,
            offset: Some(offset),
            sort_desc: true,
        }
    }

    /// Get effective page size (capped)
    pub fn effective_page_size(&self) -> usize {
        self.page_size.min(MAX_PAGE_SIZE) as usize
    }

    /// Get effective offset
    pub fn effective_offset(&self) -> usize {
        self.offset.unwrap_or(0) as usize
    }
}

/// Paginated response wrapper
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PaginatedResponse<T> {
    /// Items in this page
    pub items: Vec<T>,
    /// Total count of items (if available)
    pub total_count: Option<u64>,
    /// Cursor for next page (if cursor-based)
    pub next_cursor: Option<String>,
    /// Whether there are more items
    pub has_more: bool,
    /// Current page number (if offset-based)
    pub page: Option<u32>,
    /// Total pages (if offset-based and total known)
    pub total_pages: Option<u32>,
}

impl<T> PaginatedResponse<T> {
    /// Create an empty response
    pub fn empty() -> Self {
        Self {
            items: vec![],
            total_count: Some(0),
            next_cursor: None,
            has_more: false,
            page: Some(0),
            total_pages: Some(0),
        }
    }

    /// Create from items with pagination info
    pub fn new(items: Vec<T>, has_more: bool, next_cursor: Option<String>) -> Self {
        Self {
            items,
            total_count: None,
            next_cursor,
            has_more,
            page: None,
            total_pages: None,
        }
    }

    /// Set total count and calculate total pages
    pub fn with_total(mut self, total: u64, page_size: u32) -> Self {
        self.total_count = Some(total);
        self.total_pages = Some((total as f64 / page_size as f64).ceil() as u32);
        self
    }

    /// Set current page
    pub fn with_page(mut self, page: u32) -> Self {
        self.page = Some(page);
        self
    }
}

/// Paginate a vector of items
pub fn paginate_vec<T: Clone>(items: Vec<T>, request: &PaginationRequest) -> PaginatedResponse<T> {
    let total = items.len();
    let offset = request.effective_offset();
    let page_size = request.effective_page_size();

    if offset >= total {
        return PaginatedResponse::empty().with_total(total as u64, request.page_size);
    }

    let end = (offset + page_size).min(total);
    let page_items: Vec<T> = items[offset..end].to_vec();
    let has_more = end < total;

    let next_cursor = if has_more {
        Some(end.to_string())
    } else {
        None
    };

    PaginatedResponse::new(page_items, has_more, next_cursor)
        .with_total(total as u64, request.page_size)
        .with_page((offset / page_size) as u32)
}

/// Cursor encoding/decoding utilities
pub mod cursor {

    /// Encode a cursor from timestamp and optional ID
    pub fn encode(timestamp_micros: i64, id: Option<&str>) -> String {
        match id {
            Some(id) => format!("{}:{}", timestamp_micros, id),
            None => timestamp_micros.to_string(),
        }
    }

    /// Decode a cursor to timestamp and optional ID
    pub fn decode(cursor: &str) -> Option<(i64, Option<String>)> {
        if cursor.contains(':') {
            let parts: Vec<&str> = cursor.splitn(2, ':').collect();
            if parts.len() == 2 {
                let ts = parts[0].parse().ok()?;
                Some((ts, Some(parts[1].to_string())))
            } else {
                None
            }
        } else {
            let ts = cursor.parse().ok()?;
            Some((ts, None))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_paginate_vec() {
        let items: Vec<i32> = (1..=100).collect();

        // First page
        let req = PaginationRequest::with_offset(10, 0);
        let resp = paginate_vec(items.clone(), &req);
        assert_eq!(resp.items.len(), 10);
        assert_eq!(resp.items[0], 1);
        assert!(resp.has_more);
        assert_eq!(resp.total_count, Some(100));
        assert_eq!(resp.page, Some(0));

        // Middle page
        let req = PaginationRequest::with_offset(10, 50);
        let resp = paginate_vec(items.clone(), &req);
        assert_eq!(resp.items.len(), 10);
        assert_eq!(resp.items[0], 51);
        assert!(resp.has_more);
        assert_eq!(resp.page, Some(5));

        // Last page
        let req = PaginationRequest::with_offset(10, 90);
        let resp = paginate_vec(items.clone(), &req);
        assert_eq!(resp.items.len(), 10);
        assert_eq!(resp.items[0], 91);
        assert!(!resp.has_more);
    }

    #[test]
    fn test_cursor_encode_decode() {
        let ts = 1704825600_000_000i64;

        // Without ID
        let cursor = cursor::encode(ts, None);
        let (decoded_ts, decoded_id) = cursor::decode(&cursor).unwrap();
        assert_eq!(decoded_ts, ts);
        assert!(decoded_id.is_none());

        // With ID
        let cursor = cursor::encode(ts, Some("abc123"));
        let (decoded_ts, decoded_id) = cursor::decode(&cursor).unwrap();
        assert_eq!(decoded_ts, ts);
        assert_eq!(decoded_id, Some("abc123".to_string()));
    }

    #[test]
    fn test_page_size_cap() {
        let req = PaginationRequest::new(1000, None);
        assert_eq!(req.effective_page_size(), MAX_PAGE_SIZE as usize);
    }
}
